File: iosApp/iosApp/view/screens/search/SearchViewModel.swift
import Foundation
import shared

// 🔧 15-Jul-2026 iOS parity (Phase 2.2): state + intents for the search screen (Android
//   SearchViewModel parity). MVI: query intent -> debounced FTS5 search -> results state.
//   Results are NoteSummary items — content matches carry a snippet.
struct SearchUIState {
    var query: String = ""
    var results: [NoteSummary] = []
    var isSearching = false
    var errorMessage: String? = nil
}

// 🔧 15-Jul-2026 iOS parity (Phase 2.2): NEW — drives the (previously stub) SearchView.
//   How to use: bind `state.query` to the search field via onQueryChange(text); each change is
//   debounced 300ms (the pending task is cancelled), then adapter.searchNotes hits the shared
//   FTS5 index and results land in state on the main thread.
final class SearchViewModel: ObservableObject {
    @Published var state = SearchUIState()

    private let adapter: NotesBridgeAdapter
    private var debounceTask: Task<Void, Never>? = nil

    init(adapter: NotesBridgeAdapter = NotesBridgeAdapter()) {
        self.adapter = adapter
    }

    func onQueryChange(_ query: String) {
        state.query = query
        debounceTask?.cancel()
        guard !query.trimmingCharacters(in: .whitespaces).isEmpty else {
            state.results = []
            state.isSearching = false
            return
        }
        debounceTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: 300_000_000)   // debounce: wait for a typing pause
            guard !Task.isCancelled else { return }
            await MainActor.run { self?.search(query) }
        }
    }

    private func search(_ query: String) {
        adapter.searchNotes(query: query) { [weak self] result in
            guard let self else { return }
            switch result {
            case .loading:
                self.state.isSearching = true
            case .success(let list):
                self.state.isSearching = false
                self.state.results = list as? [NoteSummary] ?? []
            case .failure(let error):
                self.state.isSearching = false
                self.state.errorMessage = error.message
                print("searchNotes failed [\(error.code)] \(error.message)")
            case .idle:
                break
            }
        }
    }
}

File: androidApp/src/main/java/com/app/pustakam/android/screen/search/SearchViewModel.kt
package com.app.pustakam.android.screen.search

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.app.pustakam.core.model.models.response.notes.NoteSummary
import com.app.pustakam.feature.notes.domain.usecase.SearchNotesUseCase
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import com.app.pustakam.core.common.util.Result

// 🔧 15-Jul-2026 Phase 2.2: state + intents for the search screen (MVI: query intent -> debounced
//   FTS5 search -> results state). Results are NoteSummary items — content matches carry a snippet.
data class SearchUiState(
    val query: String = "",
    val results: List<NoteSummary> = emptyList(),
    val isSearching: Boolean = false,
)

// 🔧 15-Jul-2026 Phase 2.2: NEW — drives the (previously stub) SearchView.
//   How to use: collect `state`, call onQueryChange(text) from the search field; each change is
//   debounced 300ms, the previous in-flight search is cancelled, and results land in state.
class SearchViewModel : ViewModel(), KoinComponent {
    private val searchNotesUseCase by inject<SearchNotesUseCase>()
    private val _state = MutableStateFlow(SearchUiState())
    val state = _state.asStateFlow()
    private var searchJob: Job? = null

    fun onQueryChange(query: String) {
        _state.update { it.copy(query = query) }
        searchJob?.cancel()
        if (query.isBlank()) {
            _state.update { it.copy(results = emptyList(), isSearching = false) }
            return
        }
        searchJob = viewModelScope.launch {
            delay(300)   // debounce: search after the user pauses typing
            searchNotesUseCase(query).collect { result ->
                when (result) {
                    is Result.Loading -> _state.update { it.copy(isSearching = true) }
                    is Result.Success -> _state.update {
                        @Suppress("UNCHECKED_CAST")
                        it.copy(results = (result.data.data as? List<NoteSummary>).orEmpty(), isSearching = false)
                    }
                    is Result.Error -> _state.update { it.copy(isSearching = false) }
                }
            }
        }
    }
}
